noxfile.py

```diff
@@ -58,124 +77,50 @@ def tests(session: nox.Session) -> None:
-
-
-def _get_pypi_package_data(package_name):
-    json_uri = "https://pypi.org/pypi/{0}/json".format(package_name)
-    # Response format: https://warehouse.readthedocs.io/api-reference/json/#project
-    # Release metadata format: https://github.com/pypa/interoperability-peps/blob/master/pep-0426-core-metadata.rst
-    with url_lib.urlopen(json_uri) as response:
-        return json.loads(response.read())
-
-
-def _get_urls(data, version):
-    return list(
-        r["url"] for r in data["releases"][version] if _contains(r["url"], ("cp37",))
+    session.install(
+        "-t",
+        "./bundled/libs",
+        "--no-cache-dir",
+        "--implementation",
+        "py",
+        "--no-deps",
+        "--upgrade",
+        "-r",
+        "./requirements.txt",
     )
+    session.install("packaging")
 
-
-def _download_and_extract(root, url):
-    if "manylinux" in url or "macosx" in url or "win_amd64" in url:
-        root = os.getcwd() if root is None or root == "." else root
-        print(url)
-        with url_lib.urlopen(url) as response:
-            data = response.read()
-            with zipfile.ZipFile(io.BytesIO(data), "r") as wheel:
-                for zip_info in wheel.infolist():
-                    # Ignore dist info since we are merging multiple wheels
-                    if ".dist-info/" in zip_info.filename:
-                        continue
-                    print("\t" + zip_info.filename)
-                    wheel.extract(zip_info.filename, root)
-
-
-def _install_package(root, package_name, version="latest"):
-    from packaging.version import parse as version_parser
-
-    data = _get_pypi_package_data(package_name)
-
-    if version == "latest":
-        use_version = max(data["releases"].keys(), key=version_parser)
+    target = os.environ.get("VSCETARGET", "")
+    print("target:", target)
+    if "darwin" in target:
+        download_url(DEBUGPY_WHEEL_URLS["macOS"])
+    elif "win32-ia32" == target:
+        download_url(DEBUGPY_WHEEL_URLS["win32"])
+    elif "win32-x64" == target:
+        download_url(DEBUGPY_WHEEL_URLS["win64"])
+    elif "linux-x64" == target:
+        download_url(DEBUGPY_WHEEL_URLS["linux"])
     else:
-        use_version = version
-
-    for url in _get_urls(data, use_version):
-        _download_and_extract(root, url)
+        download_url(DEBUGPY_WHEEL_URLS["any"])
+
+
+def download_url(value):
+    with url_lib.urlopen(value["url"]) as response:
+        data = response.read()
+        if hashlib.sha256(data).hexdigest() != value["hash"][1]:
+            raise Exception("Failed hash verification for {}.".format(value["url"]))
+        print("Download: ", value["url"])
+        with zipfile.ZipFile(io.BytesIO(data), "r") as wheel:
+            for zip_info in wheel.infolist():
+                print("\t" + zip_info.filename)
+                wheel.extract(
+                    zip_info.filename, pathlib.Path.cwd() / "bundled" / "libs"
+                )
```

Comment:
- You don't have to hard-code the hash algorithm anymore with your earlier change
- You shouldn't ever raise `Exception`; it's too generic. You either want to create a new exception type or at least use a built-in exception that's more descriptive
- No need to recalculate the `libs/` directory every time
- You don't want to leave the socket open for any longer than necessary

```suggestion
    with url_lib.urlopen(value["url"]) as response:
        data = response.read()
    hash_algorithim, hash_digest = value["hash"]
    if hashlib.new(hash_algorithm, data).hexdigest() != hash_digest:
        raise ValueError(f"Failed hash verification for {value['url']}")
    print("Download: ", value["url"])
    with zipfile.ZipFile(io.BytesIO(data), "r") as wheel:
        libs_dir = pathlib.Path.cwd() / "bundled" / "libs"
        for zip_info in wheel.infolist():
            print("\t" + zip_info.filename)
            wheel.extract(zip_info.filename, libs_dir)
```